src/runner/test-run/index.js

```diff
@@ -0,0 +1,142 @@
+import path from 'path';
+import Promise from 'pinkie';
+import { Session } from 'testcafe-hammerhead';
+import { createDoneCommand, TYPE as COMMAND_TYPE } from './commands';
+import CLIENT_MESSAGES from '../client-messages';
+
+
+export default class TestRun extends Session {
+    constructor (test, browserConnection, screenshotCapturer, opts) {
+        var uploadsRoot = path.dirname(test.fixture.path);
+
+        super(uploadsRoot);
+
+        this.opts               = opts;
+        this.test               = test;
+        this.browserConnection  = browserConnection;
+        this.screenshotCapturer = screenshotCapturer;
+
+        this.started        = false;
+        this.pendingCommand = null;
+        this.pendingRequest = null;
+        this.pendingJsError = null;
+
+        this.errs = [];
+    }
+
+    _getPayloadScript () {
+        // TODO
+    }
+
+    _getIframePayloadScript () {
+        // TODO
+    }
+
+    async _start () {
+        this.started = true;
+
+        try {
+            await this.test.fn(this);
+        }
+        catch (err) {
+            this._fatalError(err);
+        }
+        finally {
+            this._done();
+        }
+    }
+
+    _fatalError (err) {
+        this.errs.push(err);
+    }
+
+    _done () {
+        if (this.pendingRequest)
+            this.pendingRequest.resolve(createDoneCommand());
+
+        if (this.pendingJsError)
+            this.errs.push(this.pendingJsError);
+
+        this.emit('done');
+    }
+
+    _updatePendingWaitCommand () {
+        if (this.pendingCommand && this.pendingCommand.command.type === COMMAND_TYPE.wait) {
```

Comment:
Looking at it currently I wondering why we need all this? `wait` can be just promise that resolves on timeout, we don't need response from page for it